src/Messages/BookingCodesMessage.php

```diff
@@ -0,0 +1,196 @@
+        }
+        else
+		    $this->SendNotificationMail();
+
+        remove_action( 'commonsbooking_mail_sent',array($this,'updateEmailSent'), 5 );
+        
+        return true;
+    }
+   
+    public function updateEmailSent($action,$result)
+    {
+        if($this->action != $action) return;
+            
+        if($result === true)
+        {
+            update_post_meta( (int)$this->getPostId(), \CommonsBooking\View\BookingCodes::LAST_CODES_EMAIL, time() ); 
+        }
+
+    }
+
+ 	/**
+	 * filter commonsbooking_mail_to for adding multiple to email adresses
+     * 
+     * @return array
+ 	 */
+    public function addMultiTo()
+    {
+        $to=array();
+        foreach($this->locationAdmins as $admin)
+            $to[]=sprintf( '%s <%s>', $admin->user_nicename, $admin->user_email );
+
+        return $to;
+    }
+
+ 	/**
+	 * generates iCalendar attachment with all requested booking codes
+	 *
+     * @param  array  $bookingCodes   List of BookingCode objects
+     * 
+     * @return array
+ 	 */
+    protected function getIcalAttachment($bookingCodes)
+    {
+		$calendar = new iCalendar();
+
+        foreach($bookingCodes as $bookingCode) {
+            $title=apply_filters('commonsbooking_emailcodes_icalevent_title',
+                            $bookingCode->getItemName() . " (" . $bookingCode->getCode() . ")", $bookingCode);
+            $desc=apply_filters('commonsbooking_emailcodes_icalevent_desc',
+                        sprintf( __( 'Booking code for item "%1$s": %2$s', 'commonsbooking' ), $bookingCode->getItemName(),$bookingCode->getCode()),
+                        $bookingCode);
+
+		    $calendar->addEvent( DateTimeImmutable::createFromFormat('Y-m-d',$bookingCode->getDate()), $title, $desc);
+        }
+
+        $attachment = [
+            'string' => $calendar->getCalendarData(), // String attachment data (required)
+            'filename' => "BookingCodes" . '.ics', // Name of the attachment (required)
+            'encoding' => 'base64', // File encoding (defaults to 'base64')
+            'type' => 'text/calendar', // File MIME type (if left unspecified, PHPMailer will try to work it out from the file name)
+            'disposition' => 'attachment' // Disposition to use (defaults to 'attachment')
+        ];
+
+		return $attachment;
+        
+    }
+
+ 	/**
+	 * raises mail_sent action with error info
+	 *
+     * @param  string  $title   Error msg title
+     * @param  string  $msg     Error msg content
+     * 
+     * @return bool false
+ 	 */
+      protected function raiseError(  $msg): bool {
+        do_action( 'commonsbooking_mail_sent', $this->getAction(),new \WP_Error("e-mail",$msg));
+        return false;
+    }
+
+ 	/**
+	 * returns a list of CBManager user objects
+	 *
+     * @param  array  $userIds 
+     *
+     * @return array   user objects.
+ 	 */
+      public static function getUsersFromIds($userIds): array {
```

Comment:
Das passt vielleicht besser in die UserRepository Klasse

Reply:
Ja, und ich habe die Function auch passender umbenannt in     getCBManagersByIds